### ISSUE TITLE
Crashes/blows up Chrome tab trying to Auto-Refresh

### ISSUE THREAD

**Issue by andycrum** [bdkjones/CodeKit-1-Old#144]
If for some reason a server error occurs locally while developing, CodeKit blows up Chrome auto-refreshing the page. Not sure if it happens on other browsers or not. Starts using 80%+ cpu and then finally CodeKit stops responding. Using Version 1.0 (4136).

**Comment by bdkjones** [comment]
What server error is going on? Can you send me a set of files that demonstrate this?

**Comment by andycrum** [comment]
I'm using MAMP, and it's just a backend error from Kohanna (a php framework) because of a php error.

I set CodeKit to hide and not reload all the 'pages' on save, and this fixed the problem. Not sure if it is evidence of an issue still, but this solution works for me since I am only using CodeKit for my SCSS on this project.

**Comment by colinbm** [comment]
I'm also seeing this issue, however it doesn't appear to be connected to a server error. The page is rendered correctly, but repeatedly reloads. I'm also using MAMP.

I've disabled auto-refresh, which fixes the issue for the moment. I'm using version 4742.

**Comment by AaronHolbrook** [comment]
I had the same issue pre-jan 20 update. Had to disable auto-refresh, just downloaded the new version and haven't tested it yet

**Comment by bdkjones** [comment]
Okay, I need a little more info:

1) Is it JUST CodeKit that hangs, or does Chrome hang as well until you quit CodeKit?

2) Which refresh mode are you using: "active tab" or "custom URLs"

3) If you're using "custom URLs", what have you entered in the text box that lets you specify the strings to look for? I need the VERBATIM contents of that text box.

4) If you're seeing this issue in one refresh mode, try the other and tell me if it persists.

Thanks.

**Comment by bdkjones** [comment]
Also, are you using the current RELEASE version of Chrome, or are you on the alpha or beta channel?

**Comment by bdkjones** [comment]
Please also list all plugins you have installed in Chrome. Also, if you've got any other running applications (like LiveReload), please list those.

Finally, (and this is my best guess) you're not using a Java-based editor, are you? If you are, that's the cause right there. The Java bridge in OS X has some bugs with how it saves files --- it doesn't use the standard system APIs to do it, so the kernel queues get all messed up. The primary symptom is an endless recompile of the file that you save, which would also trigger an endless refresh cycle.

The bottom line is that CodeKit is not compatible with Java-based editors. There's nothing I can do about it; the issue is on Apple's end. And since they have deprecated Java on OS X, the odds that it'll ever be fixed are exactly zero. So, it's time to move to a new editor. I recommend Sublime Text 2.

At any rate, I can not duplicate this issue on my end.

**Comment by AaronHolbrook** [comment]
I'm coding PHP and CSS/LESS - using your less preprocessor and auto-refresh. Using Google Chrome - stable channel.

I'm using Sublime Text 2. It's not an issue with the editor because I recently switched from Textmate and it was doing the same thing.

Happens regardless of what the URL is, it just dominates chrome and endlessly tries to refresh the page - have to force quit chrome or codekit.

If I turn off auto refresh everything is dandy, but the auto-refresh is one of the major things I'd like to use.

Aaron

**Comment by bdkjones** [comment]
Does it happen when you put auto-refresh into "active tab" mode?

**Comment by AaronHolbrook** [comment]
I'll keep an eye on it and let you know.

Like I said I haven't tested this new version - looks like the auto-reload got some changes.

I loaded it up and tested it both ways, so far so good.

Aaron

**Comment by bdkjones** [comment]
Are any of you guys using CodeIgnitor? Or does a log file change when the error occurs? I had another guy email me and he found that when an error occurred with CodeIgnitor code, a log file got updated, CodeKit saw that update and refreshed the browser, which led to the same error, which led to the same log file change, etc etc. Infinite loop.

**Comment by AaronHolbrook** [comment]
I am not using code igniter, just HTML, Wordpress, less, and CSS

**Comment by colinbm** [comment]
I'm using Kohana, and I think you've hit the nail on the head... I was getting a 404 error due to missing favicon.ico file - as soon as I created one and the log stopped, the browser stopped refreshing.

Kohana for some reason uses the .php extension for its log files, but presumably CodeKit should not be monitoring or reacting to .php or .log files?

**Comment by bdkjones** [comment]
Colin:

No, CodeKit _does_ react to changes to PHP files. This allows it to reload browsers when you make a change to your site's page. However, there is no way for the app to automatically know that PHP file X is a log file rather than a "real" php file that forms part of the website. Likewise, there is no way for the app to automatically know that a given change to file X was made by a framework like Kohana or CodeIgnitor rather than the user himself. All the app sees is that the file was saved and changed. 

To resolve this issue, you need to do one of these things:

1) Select the "log" php file in CodeKit and check the box that says "don't reload browser on change".

2) The first solution works great for php files that are purely log files. But if a framework is adding error messages to a real PHP file that forms part of your website, then solution 1 is not so good, as browsers won't reload when you save the file yourself after making changes. In this case, you'd need to modify the settings of whatever framework you're using so that it does not modify the PHP file with error messages. 

Aaron:

Are you sure that Wordpress is not modifying files in a similar fashion? It's been a while since I've used Wordpress, but I would _highly_ suspect that this infinite refresh loop is caused by something similar.

**Comment by AaronHolbrook** [comment]
Not to my knowledge - there are no log files or anything to that effect in wordpress and the themes I'm building.

Aaron

**Comment by andycrum** [comment]
In my case it was a php file that was displaying error messages (default devmode behavior in Kohanna I believe). Unfortunately I am no longer working on the project so I can't test it at the moment... might be able to in the next day or so.

I was using the stable version of Chrome with "localhost" in the auto-refresh custom URL field. Using TextMate2 as my editor. Hung up both the Chrome Tab and CodeKit. Let me know if you need any more information, or if I can test something specific for you.

**Comment by bdkjones** [comment]
Yea...

It's looking like this is not a real bug in CodeKit. Rather, the app is behaving exactly as it's supposed to: whenever a PHP file is changed and saved, it reloads the browsers so that you can see the changes.

The trouble is that reloading is causing frameworks like Kohanna to modify the PHP file, which triggers another reload, which triggers another Kohanna modification ad infinitum.

There is no way for CodeKit to determine what changes are "real" (i.e. implemented by the user's editing of a PHP file) and which are implemented by a given framework as "error messages." In my opinion, it's pretty lame to display errors by modifying the offending file.

So, the fix is going to be to either stop the framework from modifying your file on errors or set the file to "not reload when saved" in CodeKit.

**Comment by andycrum** [comment]
Makes sense. As I mentioned before, setting the php file to not reload on save worked just fine. Since I was just using it for SCSS I didn't need CodeKit for the php files anyway. Thanks for your help (and for CodeKit)!
